### PR TITLE
optional lan

### PR DESCRIPTION
**Pull request by ProofOfKeags**

### CODE REVIEW

File: agent/src/Handler/Apps.hs [hunk 1]
```diff
@@ -7,29 +7,29 @@
...
                                                 , asks
-                                                , Reader
-                                                , runReader
                                                 , catchError
-                                                , forkFinally
                                                 , empty
+                                                , execState
+                                                , forkFinally
+                                                , modify
+                                                , runReader
```
Comment (ProofOfKeags) [inline agent/src/Handler/Apps.hs:17]: formatter reordered shit

File: agent/src/Handler/Apps.hs [hunk 2]
```diff
@@ -7,29 +7,29 @@
...
                                                as LAsync
 import qualified Control.Concurrent.Lifted     as Lifted
-import qualified Control.Exception.Lifted      as Lifted
 import           Control.Concurrent.STM.TVar
 import           Control.Effect.Empty    hiding ( guard )
 import           Control.Effect.Labelled        ( HasLabelled
                                                 , Labelled
                                                 , runLabelled
                                                 )
+import qualified Control.Exception.Lifted      as Lifted
```
Comment (ProofOfKeags) [inline agent/src/Handler/Apps.hs:32]: formatter

File: agent/src/Handler/Apps.hs [hunk 3]
```diff
@@ -38,13 +38,13 @@ import           Data.Aeson
...
-import           Data.IORef
 import qualified Data.HashMap.Lazy             as HML
 import qualified Data.HashMap.Strict           as HM
+import           Data.IORef
 import qualified Data.List.NonEmpty            as NE
 import           Data.Singletons
-import           Data.Singletons.Prelude.Bool   ( SBool(..)
-                                                , If
+import           Data.Singletons.Prelude.Bool   ( If
+                                                , SBool(..)
```
Comment (ProofOfKeags) [inline agent/src/Handler/Apps.hs:47]: formatter

File: agent/src/Handler/Apps.hs [hunk 4]
```diff
@@ -55,8 +55,8 @@ import           Exinst
 import           Network.HTTP.Types
 import qualified Network.JSONRPC               as JSONRPC
 import           Yesod.Core.Content
-import           Yesod.Core.Json
 import           Yesod.Core.Handler      hiding ( cached )
+import           Yesod.Core.Json
```
Comment (ProofOfKeags) [inline agent/src/Handler/Apps.hs:59]: formatter

File: agent/src/Handler/Apps.hs [hunk 5]
```diff
@@ -70,9 +70,9 @@ import qualified Lib.Algebra.Domain.AppMgr     as AppMgr2
 import           Lib.Algebra.State.RegistryUrl
 import           Lib.Background
 import           Lib.Error
+import qualified Lib.External.AppManifest      as AppManifest
```
Comment (ProofOfKeags) [inline agent/src/Handler/Apps.hs:73]: formatter

File: agent/src/Handler/Apps.hs [hunk 6]
```diff
@@ -249,19 +249,28 @@ getInstalledAppsLogic = do
                 , appInstalledPreviewVersionInstalled = storeAppVersionInfoVersion
                 , appInstalledPreviewTorAddress       = Nothing
                 , appInstalledPreviewLanAddress       = Nothing
-                , appInstalledPreviewUi               = False
+                , appInstalledPreviewTorUi            = False
+                , appInstalledPreviewLanUi            = False
```
Comment (ProofOfKeags) [inline agent/src/Handler/Apps.hs:253]: here's the bifurcation of LAN/Tor UI

File: agent/src/Handler/Apps.hs [hunk 7]
```diff
@@ -249,19 +249,28 @@ getInstalledAppsLogic = do
...
             \appId (s, v, AppMgr2.InfoRes {..}) ->
-                let lanAddress = LanAddress . (".onion" `Text.replace` ".local") . unTorAddress <$> infoResTorAddress
+                let
+                    mLanAddress = do -- Maybe
+                        addrBase <- infoResTorAddress
+                        let
+                            lanConfs = mapMaybe AppManifest.portMapEntryLan
+                                $ AppManifest.appManifestPortMapping infoResManifest
+                        guard (not . null $ lanConfs)
+                        pure $ LanAddress . (".onion" `Text.replace` ".local") . unTorAddress $ addrBase
```
Comment (ProofOfKeags) [inline agent/src/Handler/Apps.hs:266]: reasons for null lan address:
1. null tor address
2. all port mappings have null lan configurations

File: agent/src/Lib/External/AppManifest.hs [hunk 8]
```diff
@@ -80,7 +81,7 @@ data AppManifest where
                     , appManifestDescShort :: Text
                     , appManifestDescLong :: Text
                     , appManifestReleaseNotes :: Text
-                    , appManifestPortMapping :: HM.HashMap Word16 Word16
+                    , appManifestPortMapping :: [PortMapEntry]
```
Comment (ProofOfKeags) [inline agent/src/Lib/External/AppManifest.hs:84]: I decided that the hashmap approach was DOA

File: agent/src/Lib/External/AppManifest.hs [hunk 9]
```diff
@@ -91,9 +92,16 @@ data AppManifest where
                     , appManifestStartAlert :: Maybe Text
                     , appManifestActions :: [Action]
                     } -> AppManifest
+deriving instance Show AppManifest
 
-uiAvailable :: AppManifest -> Bool
-uiAvailable AppManifest {..} = isJust $ HM.lookup 80 appManifestPortMapping
+torUiAvailable :: AppManifest -> Bool
+torUiAvailable AppManifest {..} = any (== 80) $ portMapEntryTor <$> appManifestPortMapping
```
Comment (ProofOfKeags) [inline agent/src/Lib/External/AppManifest.hs:98]: if any exposure of port 80 happens on tor, it has a tor UI

File: agent/src/Lib/External/AppManifest.hs [hunk 10]
```diff
@@ -91,9 +92,16 @@ data AppManifest where
...
+
+lanUiAvailable :: AppManifest -> Bool
+lanUiAvailable AppManifest {..} = any id $ fmap portMapEntryLan appManifestPortMapping <&> \case
+    Just Standard     -> True
+    Just (Custom 443) -> True
+    _                 -> False
```
Comment (ProofOfKeags) [inline agent/src/Lib/External/AppManifest.hs:104]: standard lans yield UI's, also a custom lan with 443 exposed, we may also want to add 80 here?

File: agent/src/Lib/External/AppManifest.hs [hunk 11]
```diff
@@ -114,12 +122,31 @@ instance FromJSON AppManifest where
...
 
 getAppManifest :: (MonadIO m, HasFilesystemBase sig m) => AppId -> S9ErrT m (Maybe AppManifest)
 getAppManifest appId = do
     base <- ask @"filesystemBase"
     ExceptT $ first (ManifestParseE appId) <$> liftIO
         (Yaml.decodeFileEither . toS $ (appMgrAppPath appId <> "manifest.yaml") `relativeTo` base)
+
+data LanConfiguration = Standard | Custom Word16 deriving (Eq, Show)
+instance FromJSON LanConfiguration where
+    parseJSON = liftA2 (<|>) standard custom
```
Comment (ProofOfKeags) [inline agent/src/Lib/External/AppManifest.hs:135]: try to parse as standard, failover and try custom after

File: agent/src/Lib/External/AppManifest.hs (hunk 10 above)
Comment (dr-bonez) [reply agent/src/Lib/External/AppManifest.hs:104]: Yeah, add port 80 just in case we need it

**Review by dr-bonez — CHANGES_REQUESTED** [review]
Otherwise looks good

**Review by dr-bonez — DISMISSED** [review]

**Review by dr-bonez — APPROVED** [review]

**Pull request merged**